### PR TITLE
Switch from Moment.js to the Temporal API

### PR DESCRIPTION
**Pull request by mgwalker**
# Longterm WIP

The [Temporal API](https://tc39.es/proposal-temporal/docs/index.html) is currently a TC39 stage 3 proposal, which means the API is nearly completed and is not expected to change much more before final approval. It will be a while yet before it has native implementations, but there's no reason we can't start playing with it now using the polyfill.

The goal of this WIP PR is to explore migrating every current use of Moment.js in Charlie to the Temporal API. Some of those may be harder than others, and may justify switching to a different library instead of using the Temporal API directly. Since there's no rush to make this change, a long-running PR seemed like a reasonable approach to figuring out what we should do.

The following parts of the bot rely on Moment.js and should be migrated:

- [x] angry Tock
- [x] encourage bot
- [x] federal holidays reminder
- [x] federal holidays
- [x] optimistic Tock
- [x] timezone ("Tau bot")
- [x] travel team bot
- [x] what-day-is-it (March 2020 bot)
- [x] utils/dates

This PR is a step towards addressing #295.

### CODE REVIEW

File: src/utils/dates.js [hunk 1]
```diff
@@ -1,18 +1,34 @@
...
-    .filter((h) => h.date.isAfter(now))
-    .shift();
+  return (
+    holidays
+      .allForYear(now.year)
+      .concat(holidays.allForYear(now.year + 1))
+      .map((h) => {
+        const [year, m, d] = h.dateString.split("-");
+        const month = m < 10 ? `0${m}` : m;
+        const day = d < 10 ? `0${d}` : d;
```
Comment (zachmargolis) [inline src/utils/dates.js:17]: starting in node 8, looks like [String.padStart](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart) is available?

```suggestion
        const month = String(m).padStart(2, '0');
        const day = String(d).padStart(2, '0');
```

File: src/scripts/angry-tock.js [hunk 2]
```diff
@@ -112,64 +117,61 @@ const shout = async ({ calm = false } = {}) => {
  */
 const isAngryTockDay = (now) => {
   const d = now || m();
-  return d.format("dddd") === "Monday" && !holidays.isAHoliday(d.toDate());
+  return d.dayOfWeek === 1 && !holidays.isAHoliday(dateFromZonedDateTime(m()));
```
Comment (zachmargolis) [inline src/scripts/angry-tock.js:120]: 😭  `.dayOfWeek === 1` is my least favorite part of native Date class

File: src/scripts/angry-tock.js [hunk 3]
```diff
@@ -112,64 +117,61 @@ const shout = async ({ calm = false } = {}) => {
...
+  const firstTockShoutTime = day.withPlainTime(ANGRY_TOCK_FIRST_ALERT);
+  const secondTockShoutTime = day.withPlainTime(ANGRY_TOCK_SECOND_ALERT);
+
+  let next = day;
 
   if (isAngryTockDay(day)) {
     // If today is the normal day for Angry Tock to shout...
-    if (day.isBefore(firstTockShoutTime)) {
+    if (Temporal.ZonedDateTime.compare(day, firstTockShoutTime) < 0) {
+      // day.isBefore(firstTockShoutTime)) {
```
Comment (zachmargolis) [inline src/scripts/angry-tock.js:137]: I know this is in draft, but this commented out line will go away right?

File: src/scripts/timezone.js [hunk 4]
```diff
@@ -29,7 +30,27 @@ const TIMEZONES = {
...
+      ampmStr = " pm";
+    }
+  }
+
+  const hour = h === 0 ? 12 : h;
+
+  return `${hour > 12 ? hour - 12 : hour}:${
+    minute > 10 ? "" : "0"
+  }${minute}${ampmStr}`;
+};
```
Comment (zachmargolis) [inline src/scripts/timezone.js:50]: Is there a way to use `Intl.DateTimeFormat` to implement this?

If not, and we don't want to keep moment, what about considering a smaller library like [d3-time-format](https://github.com/d3/d3-time-format) around?

File: src/scripts/angry-tock.js (hunk 3 above)
Comment (mgwalker) [reply src/scripts/angry-tock.js:137]: Ooh, good catch. Thanks!

File: src/utils/dates.js (hunk 1 above)
Comment (mgwalker) [reply src/utils/dates.js:17]: YES! Thank you! I *always* forget that `padStart` is part of the spec now.

File: src/scripts/angry-tock.js (hunk 2 above)
Comment (mgwalker) [reply src/scripts/angry-tock.js]: Yeah, it bit me *several* times. I finally gave up trying to keep it straight in my head and created an enum so I can do `.dayOfWeek === DAYS.Monday` instead.

File: src/scripts/angry-tock.js [hunk 5]
```diff
@@ -27,12 +25,19 @@ const HAPPY_TOCK_REPORT_TO = (process.env.HAPPY_TOCK_REPORT_TO || "#18f").split(
...
+
 /**
  * Get the current time in the configured timezone.
- * @returns {Moment} A moment object representing the current time in the
- *   configured timezone.
+ * @returns {Temporal.ZonedDateTime} A Temporal object representing the current
+ *   time in the configured timezone.
  */
-const m = () => moment.tz(ANGRY_TOCK_TIMEZONE);
+const m = () =>
```
Comment (zachmargolis) [inline src/scripts/angry-tock.js:36]: Maybe rename this to `t` or even something verbose like `currentZonedDateTime()`?

```suggestion
const currenetZonedDateTime = () =>
```

File: src/scripts/angry-tock.js [hunk 6]
```diff
@@ -107,69 +112,68 @@ const shout = async ({ calm = false } = {}) => {
...
+        dateFromZonedDateTime(firstTockShoutTime),
+        async () => {
+          await shout({ calm: true });
+          setTimeout(() => scheduleNextShoutingMatch(), 1000);
+        }
+      );
     }
-    if (day.isBefore(secondTockShoutTime)) {
+    if (Temporal.ZonedDateTime.compare(day, secondTockShoutTime) < 0) {
+      // if (day.isBefore(secondTockShoutTime)) {
```
Comment (zachmargolis) [inline src/scripts/angry-tock.js:151]: another extra commented out line

File: src/scripts/angry-tock.js (hunk 5 above)
Comment (mgwalker) [reply src/scripts/angry-tock.js]: Yes! I put this on my mental checklist of things to come back and do and then forgot. But your comment also reminded me that I'd made this functionality available from the dates utility module, so I switched to using that. 😄

**Comment by mgwalker** [comment]
Closing as ancient.

**Pull request closed without merging**